parser/unparser.go

```diff
@@ -157,23 +157,39 @@ func (un *unparser) visitCallConditional(expr *exprpb.Expr) error {
 	c := expr.GetCallExpr()
 	args := c.GetArgs()
 	// add parens if operand is a conditional itself.
+	// otherwise defer decision to visitMaybeParenExpr.
 	nested := isSamePrecedence(operators.Conditional, args[0])
```

Comment:
The precedence and paren expr checking functions are more or less doing the same thing, i.e. is the argument an operator with more than one argument? Instead of checking for precedence, I would instead create a function called `isComplexOperator`:

```go
func isComplexOperator(arg *exprpb.Expr) bool {
  call := arg.GetCallExpr()
  // If the arg is not a call with more than one arg, return false.
  if call == nil || len(call.GetArgs()) < 2 {
    return false
  }
  // Otherwise, return whether the function name is listed as an operator.
  _, isOp := operators.FindReverse(call.GetFunction())
  return isOp
}
```

Then I would recommend writing the code as 

```go
nested := isComplexOperator(args[0])
err = un.visitMaybeNested(args[0], nested)
```

Reply:
This function by itself is insufficient, it generates invalid expressions on conditional, since it also has to take non-call functions into account. I have added an additional commit which could go into the direction of what you proposed, but extends it. I'm not sure whether this really improves clarity.

If you run just the version with `isComplexOperator`, you will see that the tests no longer pass, since `a ? (b ? c : d) : e` will be unparsed into the invalid expression `a ? b ? c : d : e`.

Reply:
Ah my apologies, I forgot that the `FindReverse` didn't account for ternary operators. Probably some cleanup is due there to make its purpose clearer.

Instead, you'd write the following to get the desired effect:

```go
func isComplexOperator(arg *exprpb.Expr) bool {
	call := arg.GetCallExpr()
	// If the arg is not a call with more than one arg, return false.
	if call == nil || len(call.GetArgs()) < 2 {
		return false
	}
	// Otherwise, return whether the function name is listed as an operator  with any precedence.
	return operators.Precedence(call.GetFunction()) != 0
}
```

I guess my point is that the current function for `isComplexOperator` is more complex than it needs to be.

Reply:
Made this more clear with the new commits, uses the same structure as the C++ decompiler now and added several more test cases which were problematic before.